unixfs/src/dir/walk.rs

```diff
@@ -0,0 +1,777 @@
+                self.set_path(name, depth);
+                self.metadata = metadata;
+            },
+            Bucket(_)
+            | Directory(_)
+            | File(_, None)
+            | Symlink(_) => {
+                self.kind = File(cid, step);
+                self.set_path(name, depth);
+                self.metadata = metadata;
+            }
+            ref x => todo!("file from {:?}", x),
+        }
+    }
+
+    fn as_symlink(
+        &mut self,
+        cid: Cid,
+        name: &str,
+        depth: usize,
+        metadata: FileMetadata
+    ) {
+        use InnerKind::*;
+        match self.kind {
+            RootDirectory => {
+                self.kind = Symlink(cid);
+                self.set_path(name, depth);
+                self.metadata = metadata;
+            },
+            Bucket(_)
+            | Directory(_)
+            | File(_, None)
+            | Symlink(_) => {
+                self.kind = Symlink(cid);
+                self.set_path(name, depth);
+                self.metadata = metadata;
+            },
+            ref x => todo!("symlink from {:?}", x),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Walk<'a> {
+    /// Walk was started on a file node which we don't know any name for
+    SingleBlockFile {
+        bytes: &'a [u8],
+        metadata: FileMetadata,
+    },
+    Symlink {
+        metadata: FileMetadata,
+        target: &'a [u8],
+    },
+    /// Walk was started on a file which can be iterated further with the given FileVisit
+    // TODO: not sure why not expose this through the continue_block as well? the path will just be
+    // empty all the time.
+    MultiBlockFile {
+        metadata: FileMetadata,
+        visit: FileVisit,
+    },
+    EmptyDirectory {
+        metadata: FileMetadata,
+    },
+    /// Walk was started on a directory
+    Walker(Walker),
+}
+
+// Wrapper to hide the state
+#[derive(Debug)]
+pub struct Item {
+    state: State,
+}
+
+impl From<State> for Item {
+    fn from(state: State) -> Self {
+        Item {
+            state
+        }
+    }
+}
+
+impl Item {
+    // TODO: add path(&self) -> &Path
+
+    pub fn as_entry(&self) -> Entry<'_> {
+        match &self.state {
+            State::Unfinished(w) => w.as_entry(),
+            State::Last(w) => w.as_entry(),
+        }
+    }
+
+    pub fn into_inner(self) -> Option<Walker> {
+        match self.state {
+            State::Unfinished(w) => Some(w),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum State {
+    Unfinished(Walker),
+    Last(InnerEntry),
+}
+
+#[derive(Debug)]
+pub enum ContinuedWalk<'a> {
+    // should this be just the description of what we just walked into?
+    File(FileSegment<'a>, Item),
+    Directory(Item),
+    Symlink(&'a [u8], Item),
+}
+
+#[derive(Debug)]
+pub struct Skipped(State);
+
+#[derive(Debug)]
+pub struct FileSegment<'a> {
+    bytes: &'a [u8],
+    first_block: bool,
+    last_block: bool,
+}
```

Comment:
this needs to contain the total file size and something like that ... or maybe it could be over at FileMetadata, except that it's present for Files and Symlinks but not for directories and hamt shards.

Reply:
Added already but will not update this PR.